programs/bpf_loader/src/lib.rs

```diff
@@ -1064,6 +1065,36 @@ fn process_loader_upgradeable_instruction(
                 return Err(InstructionError::InvalidArgument);
             }
 
+            let program_account = instruction_context
+                .try_borrow_instruction_account(transaction_context, PROGRAM_ACCOUNT_INDEX)?;
+            if !program_account.is_writable() {
+                ic_logger_msg!(log_collector, "Program account is not writable");
+                return Err(InstructionError::InvalidArgument);
+            }
+            if program_account.get_owner() != program_id {
+                ic_logger_msg!(log_collector, "Program account not owned by loader");
+                return Err(InstructionError::InvalidAccountOwner);
+            }
+            match program_account.get_state()? {
+                UpgradeableLoaderState::Program {
+                    programdata_address,
+                } => {
+                    if programdata_address != programdata_key {
+                        ic_logger_msg!(
+                            log_collector,
+                            "Program account does not match ProgramData account"
+                        );
+                        return Err(InstructionError::InvalidArgument);
+                    }
+
+                    drop(program_account);
+                }
+                _ => {
+                    ic_logger_msg!(log_collector, "Invalid Program account");
+                    return Err(InstructionError::InvalidAccountData);
+                }
+            }
```

Comment:
Can we move `drop(program_account);` here?

Reply:
Yeah, that's more clear. Moved.